designs/Transitive-Dependencies.md

```diff
@@ -0,0 +1,150 @@
+
+Knowing how a package is being included in a project helps a developer diagnose many problems. It also provides them with an opportunity to understand complex dependency chains they may have within their projects. 
+There is not a great solution that exists to understand the nature of top-level packages and their transitive dependencies.
+The solution explorer in Visual Studio does the best job at allowing a user to dive into each NuGet package & see all transitive dependencies from the top-level dependency.
+
+![](../resources/TransitiveDependencies/SolutionView.png)
+
+The dotnet CLI does not provide the insight into “why” a transitive dependency is listed, although it does list everything that has been resolved.
+
+![](../resources/TransitiveDependencies/DotNetCLI.png)
+
+The Visual Studio Package Manager UI does not support the concept of showing transitive packages in any fashion.
+
+![](../resources/TransitiveDependencies/VSPMUI.png)
+
+## Success: How do we know if we have solved this problem?
+
+By measuring the number of detractors regarding transitive dependencies & promoters of this new experience in future NuGet client surveys.
+
+We will add telemetry to categorize when a top-level package is installed, and whether it was done in the context of promoting a transitive package to a top-level one.
+
+## Audience: Who are we building this for?
+
+We are building this for .NET users, both old and new, to be successful with understanding the types of NuGet dependencies in their project(s) and solution(s).
+
+## What: What does this look like in the product?
+
+![](../resources/TransitiveDependencies/TransitiveVSPMUI.png)
+
+## Context and Scope
+
+-	https://github.com/NuGet/Home/issues/5887 
+-	https://developercommunity.visualstudio.com/content/idea/538119/add-dependency-tree-view-for-packagereferences.html 
+
+## Minimal Requirements
+
+-	Visual list of transitive dependencies.
+-	Ability to see a transitive dependency package details.
+-	Ability to determine where a transitive dependency originated.
+-	Ability to promote a transitive dependency to a top-level dependency.
+
+## Scenarios and User Experience
+
+### CLI
+
+We should consider a minor tweak to the existing experience of “dotnet list package --include-transitive” to provide the user with a sense of where the package came from. This could be a new column next to “Resolved” which says “Transitively Referenced” or “Referenced” and has a list of the top-level packages that requested the dependency.
+
+![](../resources/TransitiveDependencies/TransitiveDotNetCLI.png)
+
+### Visual Studio
+
+Within Visual Studio, there will be a new panel within the Visual Studio Package Manager UI named “Transitive Packages” in which all transitive packages will be displayed to the user. For the sake of not confusing the user, there will additionally be a header titled “Top-level Packages” where the current experience is today for top-level packages.
```

Comment:
Is this spec committing to this specific implementation within PM UI? If not, the wording should be changed as such. Also consider whether there will be an iterative approach to bringing transitives to the UI, and what those phases might look like.

I think the collapsing panel is a fine paradigm, but did we consider alternatives for tops vs transitives? 
Perhaps a filter option to "Show transitives" would work just as well? 
The description in https://github.com/NuGet/Home/issues/5887 shows a screenshot with a highlight around a package that's "installed explicitly"/top. This made me think that a visual indicator (color/etc) and screen-reader property may work just as well. And much like our `show prerelease`, they can be filtered out.

What do we gain from a collapsible panel? If there's always going to be on the order of N=10 top-levels, and N=100's of Transitives, it's hard to envision why I would want the collapsing behavior.

Reply:
Yes this spec is committing to this implementation as there was really no suggested alternative & people generally agreed this was a good approach. If we don't feel that way today, we should reiterate.

This work by itself is phased from the package vulnerability work. Because of the scope of this work, there's really no phased approach to this specific design.

The reason there is a collapsing concept is to allow users to have their screen real estate back if they do not care for transitive dependencies and would like to keep them collapsed while doing top-level management tasks.

The decision for two separate lists was primarily driven by many different constructs such as the existing dotnet CLI behavior for showing transitives, and the many eco systems that show top-level dependencies & transitive dependencies in separate lists. (gradle, ant, etc). Others show dependency trees, which was not designed previously and thus we're committing to a similar experience already completed in dotnet.

The jobs to be done are to look at top-level, transitive level, or both. There are many reasons one may look at each level & having collapsable controls helps users decide how they'd like their screen to be shown. Otherwise yes, we could do a filter, but the purpose was to differentiate the types of references clearly and not have them all in one list with a UI affordance to say "transitive" or "top-level" although you might be able to filter/order by that.